src/main/java/uk/gov/justice/hmpps/prison/service/BedAssignmentHistoryService.java

```diff
@@ -78,10 +80,24 @@ public BedAssignmentHistoryPK add(final Long bookingId, final Long livingUnitId,
             .collect(Collectors.toList());
     }
 
-    public List<BedAssignment> getBedAssignmentsHistoryByDate(final LocalDate assignmentDate) {
-        return repository
-            .findBedAssignmentHistoriesByAssignmentDate(assignmentDate)
+    @VerifyAgencyAccess
+    public List<BedAssignment> getBedAssignmentsHistoryByDateForAgency(final String agencyId, final LocalDate assignmentDate) {
+        final var assignments = new ArrayList<>(repository.findBedAssignmentHistoriesByAssignmentDate(assignmentDate));
+
+        final var livingUnitIds = assignments.stream()
+            .map(BedAssignmentHistory::getLivingUnitId)
+            .collect(Collectors.toSet());
+
+        final var agencyLivingUnitIds = livingUnitIds
+            .stream()
+            .map(locationRepository::findOneByLocationId)
+            .filter(agencyInternalLocation -> agencyInternalLocation.map(loc -> loc.getAgencyId().equals(agencyId)).orElse(false))
+            .map(agencyInternalLocation -> agencyInternalLocation.get().getLocationId())
+            .collect(Collectors.toSet());
+
+        return assignments
             .stream()
+            .filter(assignment -> agencyLivingUnitIds.contains(assignment.getLivingUnitId()))
```

Comment:
It might be simpler to just do the filtering here (get agencyInternalLocation then compare to agencyId) and we can get rid of the above pre-processing?